soot-infoflow-android/src/soot/jimple/infoflow/android/source/AndroidSourceSinkManager.java

```diff
@@ -906,6 +906,32 @@ public void setAppPackageName(String appPackageName) {
 		this.appPackageName = appPackageName;
 	}
 
+	private SootMethod grabMethod(String sootClassName,String subSignature)
+	{
+		Scene scene=Scene.v();
+		if(!scene.containsClass(sootClassName))
```

Comment:
Use Scene.v().getSootClassUnsafe() instead of first calling contains and then get

Reply:
It gives me the following exception on the line "sootMethods = sootClass.getMethods();" : 

This operation requires resolving level SIGNATURES but java.util.AbstractMap$SimpleImmutableEntry is at resolving level DANGLING
If you are extending Soot, try to add the following call before calling soot.Main.main(..):
Scene.v().addBasicClass(java.util.AbstractMap$SimpleImmutableEntry,SIGNATURES);
Otherwise, try whole-program mode (-w).

Do you have any idea why?

Reply:
That's odd. Can you show the call stack of the exception?

Reply:
at soot.SootClass.checkLevelIgnoreResolving(SootClass.java:182)
	at soot.SootClass.checkLevel(SootClass.java:162)
	at soot.SootClass.getMethods(SootClass.java:501)
	at soot.jimple.infoflow.android.source.AndroidSourceSinkManager.grabMethodWithoutReturn(AndroidSourceSinkManager.java:933)
	at soot.jimple.infoflow.android.source.AndroidSourceSinkManager.initialize(AndroidSourceSinkManager.java:973)
	at soot.jimple.infoflow.Infoflow.runAnalysis(Infoflow.java:268)
	at soot.jimple.infoflow.Infoflow.runAnalysis(Infoflow.java:232)
	at soot.jimple.infoflow.android.SetupApplication$InPlaceInfoflow.runAnalysis(SetupApplication.java:1172)
	at soot.jimple.infoflow.android.SetupApplication.runInfoflow(SetupApplication.java:1389)
	at soot.jimple.infoflow.android.SetupApplication.runInfoflow(SetupApplication.java:1293)
	at soot.jimple.infoflow.cmd.MainClass.run(MainClass.java:236)
	at soot.jimple.infoflow.cmd.MainClass.main(MainClass.java:196)

Reply:
Very strange. Does a check if (!sootClass.isPhantom()) solve the problem? In the worst case, you can check the resolving level before accessing the methods, although that would not be very clean style.

Reply:
It solved by adding if(sootClass.resolvingLevel()!=0), could you tell me what isPhantom and resolving level mean in brief to fully understand the code?

Reply:
@StevenArzt

Reply:
Please do not use the numbers, use the constants from the SootClass instead. 0 corresponds to DANGLING. The resolving level indicates  how much of the class has already been loaded by Soot. Due to performance reasons, Soot does not fully load all classes from the classpath. It determines which classes are necessary and whether method bodies are required, or whether the method signatures alone are sufficient, or whether the hierarchy of the class must be loaded.

A phantom class is a class that has been referenced by some other code (i.e., a method invocation), but is not available on the classpath. This allows Soot to model external dependencies, without requiring everything to be part of the analysis.